src/main/java/io/codelex/flightplanner/repository/FlightRepository.java

```diff
@@ -24,12 +25,14 @@ public Flight getFlight(Long id) {
                       .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
     }
 
-    public synchronized Flight addFlight(Flight flight) {
+    @Synchronized
+    public Flight addFlight(Flight flight) {
```

Comment:
Te varētu būt āķis - manuprāt nestrādās, ja addFlight un exists būs atsevišķi katra savā sinhronizētajā metodē.
Iemesls ir salīdzinoši vienkāršs.
Iedomājies situāciju, kad paralēli cenšamies pievienot divus vienādus Flight. Abi thread A un B paralēli izsauc service.add metodi. Tad nonāk līdz flightExists, kura ir sinhronizēta, tāpēc A to izpilda vispirms un turpina savu darbu, pēc tam to pilda B. Un te jau ir problēma - ja B paspēs izpildīt flightExists metodi pirms A ir reāli paspējis to Flight pievienot - tad arī B saņems atbildi, ka tāds flight neeksistē un abi thread pievienos vienu un to pašu flight.
Tas ir līdzīgi, kā ar to `i++` darbību, kura faktiski ir nevis viena operācija bet 3 - get, add un set. Un ja sinhronizēta būtu katra atsevišķi, tad tas nepalīdz - sinhronizētām jābūt visām trim kopā. Vai arī pareizāk - tai ir jābūt atomic darbībai.

Reply:
Piekrītu, Service slānī arī nepieciešama sinhronizācija, jo tur tomēr ir četri izsaukum uz repozitorijiem.